artemis-commons/src/main/java/org/apache/activemq/artemis/utils/collections/LinkedListImpl.java

```diff
@@ -203,20 +223,31 @@ public void addSorted(E e) {
          // This would be an optimization for our usage.
          // avoiding scanning the entire List just to add at the end, so we compare the end first.
          if (comparator.compare(tail.val(), e) >= 0) {
+            logger.trace("addTail as compare is after tail {}", e);
             addTail(e);
             return;
          }
 
-         Node<E> fetching = head.next;
-         while (fetching.next != null) {
-            int compareNext = comparator.compare(fetching.next.val(), e);
-            if (compareNext <= 0) {
-               addAfter(fetching, e);
-               return;
+         if (lastAdd != null) { // as an optimization we check against the last add rather than always scan.
```

Comment:
Should probably copy the value rather than accessing it repeatedly given it is volatile, and also makes it less brittle in terms of its not-thread-safe'ty (though obviously the whole list isnt).